contracts/03-tamagotchi-nft/src/lib.rs

```diff
@@ -1,19 +1,131 @@
 #![no_std]
 
 #[allow(unused_imports)]
-use gstd::prelude::*;
+use gstd::{exec, msg, prelude::*};
+use tamagotchi_nft_io::{Tamagotchi, TmgAction, TmgEvent};
+
+static mut TAMAGOTCHI_STATE: Option<Tamagotchi> = None;
+
+const HUNGER_PER_BLOCK: u64 = 1;
+const BOREDOM_PER_BLOCK: u64 = 2;
+const ENERGY_PER_BLOCK: u64 = 2;
+const FILL_PER_FEED: u64 = 1000;
+const FILL_PER_ENTERTAINMENT: u64 = 1000;
+const FILL_PER_SLEEP: u64 = 1000;
 
 #[no_mangle]
 extern fn init() {
-    // TODO: 0️⃣ Copy the `init` function from the previous lesson and push changes to the master branch
+    let name: String = msg::load().expect("Invalid initial name");
+
+    let date_of_birth = exec::block_timestamp();
+
+    let tamagotchi = Tamagotchi {
+        name,
+        date_of_birth,
+        owner: msg::source(),
+        fed: 1000,
+        fed_block: 0,
+        entertained: 1000,
+        entertained_block: 0,
+        slept: 1000,
+        slept_block: 0,
+        approved_account: None,
+    };
+
+    save_tamagotchi_state(tamagotchi);
 }
 
 #[no_mangle]
 extern fn handle() {
-    // TODO: 0️⃣ Copy the `handle` function from the previous lesson and push changes to the master branch
+    let action: TmgAction = msg::load().expect("Failed to decode TmgAction");
+
+    let current_block_height = exec::block_height() as u64;
+    let tamagotchi = unsafe {
+        TAMAGOTCHI_STATE
+            .as_mut()
+            .expect("The contract is not initialized")
+    };
+
+    if msg::source() != tamagotchi.owner && Some(msg::source()) != tamagotchi.approved_account {
+        panic!("Unauthorized");
+    }
```

Comment:
This check must be done for `Transfer`
Because now only admin and approved account are able to request the Name, Age